### PR TITLE
Implement ParetoNBD with covariates

### PR DESCRIPTION
**Pull request by ricardoV94**
## Related Issue

- [x] Supersedes and closes #463
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [ ] MMM
- [x] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--545.org.readthedocs.build/en/545/

### CODE REVIEW

**Comment by ColtAllen** [comment]
I think an updated UML diagram would be a helpful addition to CONTRIBUTING.MD to better understand all these new class inheritances:

image

 Here are some instructions on how to create one:

https://github.com/pymc-labs/pymc-marketing/issues/178

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 1]
```diff
@@ -750,3 +561,491 @@ def distribution_customer_population(
...
+        List containing column names of covariates for customer purchase rates.
+    dropout_covariate_cols: list, optional
+        List containing column names of covariates for customer dropouts.
+    model_config: dict, optional
+        Dictionary containing model parameters:
+            * `r_prior`: Shape parameter of time between purchases; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
+            * `alpha_prior`: Scale parameter of time between purchases; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `s_prior`: Shape parameter of time until dropout; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
+            * `beta_prior`: Scale parameter of time until dropout; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `purchase_covariates_prior`: Coefficients for purchase rate covariates; defaults to `pymc.Normal.dist(mu=0, sigma=1, n=len(purchase_covariates_cols)`
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/pareto_nbd.py:743]: Why the change from `StudentT` for default priors?

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 2]
```diff
@@ -211,59 +215,152 @@ def default_model_config(self) -> Dict[str, Dict]:
             "alpha_prior": {"dist": "Weibull", "kwargs": {"alpha": 2, "beta": 10}},
             "s_prior": {"dist": "Weibull", "kwargs": {"alpha": 2, "beta": 1}},
             "beta_prior": {"dist": "Weibull", "kwargs": {"alpha": 2, "beta": 10}},
+            # TODO: Cauchy prior is a bit too extreme? Probably not needed if parameters are standardized
+            "dropout_covariates_prior": {
+                "dist": "StudentT",
+                "kwargs": {"nu": 1},
```
Comment (BBDS-Colt) [inline pymc_marketing/clv/models/pareto_nbd.py:221]: Did you say `"kwargs": {"nu": 2},` is needed for data to be centered at 0?  For these covariate coefficients to be interpretable, covariates need to be scaled between 0 and 1, with a prior supporting negative values. `StudentT` slows down model fitting considerably, so I'm open to alternatives provided the distribution is centered.

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 3]
```diff
@@ -211,59 +215,152 @@ def default_model_config(self) -> Dict[str, Dict]:
...
+            },
+            "purchase_covariates_prior": {
+                "dist": "StudentT",
+                "kwargs": {"nu": 1},
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/pareto_nbd.py:225]: Did you say "kwargs": {"nu": 2}, is needed for data to be centered at 0? For these covariate coefficients to be interpretable, covariates need to be scaled between 0 and 1, with a prior supporting negative values. `StudentT slows down model fitting considerably, so I'm open to alternatives.

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 4]
```diff
@@ -389,43 +521,6 @@ def expected_purchases(
             "chain", "draw", "customer_id", missing_dims="ignore"
         )
 
-    def expected_purchases_new_customer(
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/pareto_nbd.py:392]: It's worth noting `*new_customer*` methods could support user-specified covariates for what-if and sensitivity analysis, but it's a very niche use case.

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 5]
```diff
@@ -637,28 +774,60 @@ def _distribution_new_customers(
...
+                if data:
+                    new_data = {}
+                    if self.purchase_covariate_cols:
+                        new_data["purchase_data"] = data[self.purchase_covariate_cols]
+                    if self.dropout_covariate_cols:
+                        new_data["dropout_data"] = data[self.dropout_covariate_cols]
+                    pm.set_data(new_data, coords={"customer_id": data["customer_id"]})
+                else:
+                    raise ValueError(
+                        "Dataset with new covariates must be provided when sampling new customers"
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/pareto_nbd.py:810]: With covariates, each customer will have their own `alpha` and `beta` params, so the Xarray dims, will look like `alpha[customer_id0], alpha[customer_id1]`, etc. For a future PR, I wonder if a plotting function like this is doable:

image

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 6]
```diff
@@ -667,86 +836,92 @@ def _distribution_new_customers(
...
+    def distribution_customer_population(
         self,
+        T: Union[None, np.ndarray, pd.Series] = None,
         random_seed: Optional[RandomState] = None,
     ) -> xarray.Dataset:
-        """Sample from the Gamma distribution representing dropout times for new customers.
+        """Pareto/NBD process representing purchases across the customer population.
 
-        This is the duration of time a new customer is active before churning, or dropping out.
+        This is the distribution of purchase frequencies given 'T' observation periods for each customer.
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/pareto_nbd.py:856]: If this function could support covariates, we can recreate the frequency PPC plot you're familiar with for covariate model evaluation with covariates. It's already conditional on a `T` array of observations.

**Comment by ColtAllen** [comment]
Code is certainly cleaner than the IF statement jungle of the PR I posted haha. Best way I can contribute is to create a child branch and work in parallel in the dev notebook for user-oriented feedback.

**Comment by ricardoV94** [comment]
> Code is certainly cleaner than the IF statement jungle of the PR I posted haha. Best way I can contribute is to create a child branch and work in parallel in the dev notebook for user-oriented feedback.

It's always easier to refactor than write the first pass. The notebook would help a lot! 

I'll be testing the covariate model today, and fix stuff I've certainly broken

File: tests/clv/models/test_pareto_nbd.py [hunk 7]
```diff
@@ -394,15 +372,262 @@ def test_save_load_pareto_nbd(self):
...
+        ] -= 1  # more likely to purchase (negative coefficient)
+        test_data_alt[
+            "dropout_cov"
+        ] -= 1  # less likely to drop out (positive coefficient)
+        res_high_transaction = model.predictive_distribution(test_data_alt).mean(
+            ("chain", "draw")
+        )
+        # FIXME: Test is failing for the first two conditions, signs seem swapped?
+        assert (res_zero["purchase_rate"] < res_high_transaction["purchase_rate"]).all()
+        assert (res_zero["dropout"] > res_high_transaction["dropout"]).all()
```
Comment (ricardoV94) [inline tests/clv/models/test_pareto_nbd.py:615]: @ColtAllen These two checks fail while the two below pass. Are we perhaps taking draws of the reciprocal of these quantities?

Comment (ColtAllen) [reply tests/clv/models/test_pareto_nbd.py:615]: Yes; these two lines should not be reciprocals:

https://github.com/ricardoV94/pymc-marketing/blob/6ba6ee1e7db2a93467beea974079fb0039cafca8/pymc_marketing/clv/models/pareto_nbd.py#L795

https://github.com/ricardoV94/pymc-marketing/blob/6ba6ee1e7db2a93467beea974079fb0039cafca8/pymc_marketing/clv/models/pareto_nbd.py#L801

**Comment by ricardoV94** [comment]
@ColtAllen finally pushed my last changes. I broke the pre-existing methods API very much, but we can:

1. Go back and make it like it looked before, plus optional argument for covariates in all methods
2. Use a new API but support old API with deprecation warnings
3. Something else. I don't like asking for a whole dataframe in cases where the user is only changing one variable, but for covariates it makes sense to request a dataframe anyway, so it makes sense that other variable also comes in the same dataframe

I'm in favor of 2.

Finally, the slow `TestParetoNBDModelWithCovariates::test_inference` is failing badly. I used synthetic data from fixed true parameters (so I could control the covariate coefficients), but it is not recovering those at all. Should I use different true parameter values?

File: tests/clv/models/test_pareto_nbd.py (hunk 7 above)
Comment (ricardoV94) [reply tests/clv/models/test_pareto_nbd.py]: Fixed

**Comment by ColtAllen** [comment]
> @ColtAllen finally pushed my last changes. I broke the pre-existing methods API very much, but we can:
> 
> 1. Go back and make it like it looked before, plus optional argument for covariates in all methods
> 2. Use a new API but support old API with deprecation warnings
> 3. Something else. I don't like asking for a whole dataframe in cases where the user is only changing one variable, but for covariates it makes sense to request a dataframe anyway, so it makes sense that other variable also comes in the same dataframe
> 
> I'm in favor of 2.

Yes, let's do option 2.

> Finally, the slow `TestParetoNBDModelWithCovariates::test_inference` is failing badly. I used synthetic data from fixed true parameters (so I could control the covariate coefficients), but it is not recovering those at all. Should I use different true parameter values?

Yes - the true parameters you're using are for a model fit without covariates. When I fit a covariate model in the dev notebook of my previous PR, all parameter values changed.

Also, only the covariates were synthesized, and were tacked onto an existing dataset. If we're to test with parameter recovery, we'll need to generate a full dataset from the parameters used for testing.

**Comment by ColtAllen** [comment]
> No, I'm generating new observations (recency and frequency) in that test as well

Buy why is only a single prior predictive sample being taken?

**Comment by ricardoV94** [comment]
> > No, I'm generating new observations (recency and frequency) in that test as well
> 
> Buy why is only a single prior predictive sample being taken?

One draw is enough, it contains an observation for each synthetic customer

**Comment by ColtAllen** [comment]
>  No, I'm generating new observations (recency and frequency) in that test as well
> One draw is enough, it contains an observation for each synthetic customer

I see. The new frequency column name is misspelled though, so it's testing on the original column rather than what was generated for the test.

**Comment by ColtAllen** [comment]
Is the user API still a work in progress? All predictive methods require arguments to be converted into dataframe columns beforehand, which seems rather cumbersome.

**Review by ColtAllen — APPROVED** [review]
